File: examples/node/lifecycle/handler.ts
import type { HandlerResult, StartTaskMessage, TaskContext } from '@blocks-network/sdk';

/**
 * lifecycle handler — teaches the boundary between task-lifecycle ops the
 * framework backs first-class and ops a provider composes from primitives.
 *
 *   - cancel  (built-in)            cooperative: the SDK aborts ctx.cancelSignal
 *                                   and the loop stops → terminal `canceled`.
 *   - pause   (provider-composed)   NOT a handler hook — the framework's
 *   - resume  (provider-composed)   PauseTask/ResumeTask only publish a status
 *                                   event and never reach handler code. So we
 *                                   build real work-suspension ourselves: read
 *                                   { ctrl: 'pause' | 'resume' } on a bidi
 *                                   control stream and park the work loop.
 *   - retry   (consumer-composed)   NOT done here — task state is in-memory
 *                                   only (SDK_CONTRACT §17), so retry is a
 *                                   consumer resubmit. This handler stays
 *                                   idempotent and honors a `failOnce` flag so
 *                                   the consumer can drive a failed→completed
 *                                   retry against a stateless agent.
 *
 * Task-kind split:
 *   - pipe:    long-running work loop with pause/resume/cancel (no auto-terminal).
 *   - request: single-shot; throws when `failOnce` is set (drives the retry demo).
 */
export default async function handler(
  task: StartTaskMessage,
  ctx?: TaskContext,
): Promise<HandlerResult> {
  const log = (msg: string) => console.log(`[lifecycle] ${msg}`);
  const params = parseParams(task.requestParts);
  const isPipe = !task.taskKind || task.taskKind === 'pipe';

  // --- request path: fail-once retry demo -------------------------------
  // A stateless throw. The consumer sends failOnce on attempt 1 (→ terminal
  // `failed`) and resubmits with a fresh idempotencyKey and no flag (→
  // `completed`). No persisted "have I run before?" state is needed.
  if (!isPipe) {
    if (params.failOnce) {
      log(`Task ${task.taskId}: failOnce set — throwing to produce a failed terminal`);
      throw new Error('lifecycle: simulated first-attempt failure (failOnce)');
    }
    log(`Task ${task.taskId}: request completed`);
    return {
      artifacts: [{
        data: JSON.stringify({ kind: 'request', completionReason: 'completed' }, null, 2),
        mimeType: 'application/json',
      }],
    };
  }

  // --- pipe path: pause / resume / cancel work loop ---------------------
  // Throw (not return) on a missing stream: a pipe handler's voluntary return
  // publishes NO terminal, so returning an error artifact would leave the task
  // lingering until duration expiry and then terminal as `completed`. A throw
  // publishes an immediate, correct `failed` — matching the request path.
  if (!ctx || !ctx.hasStream) {
    throw new Error('lifecycle pipe task requires a negotiated stream');
  }

  log(`Task ${task.taskId}: starting work loop (up to ${params.ticks} ticks)`);
  const control = await ctx.createStream({ declaredStream: 'control', direction: 'bidirectional', format: 'events' });

  // Read control messages on a detached loop so the work loop below is never
  // blocked waiting on input. `paused` is the single source of truth the work
  // loop parks on; `resumed` records that we ever came back from a pause (for
  // the summary). The reader unwinds when the stream is torn down.
  const state = { paused: false, resumed: false };
  const reader = (async () => {
    try {
      for await (const ev of control.events<{ ctrl?: string }>()) {
        const ctrl = typeof ev === 'object' && ev ? ev.ctrl : undefined;
        if (ctrl === 'pause' && !state.paused) {
          state.paused = true;
          log('control: pause — parking work loop');
          ctx.reportStatus('paused');
        } else if (ctrl === 'resume' && state.paused) {
          state.paused = false;
          state.resumed = true;
          log('control: resume — continuing work loop');
          ctx.reportStatus('running');
        }
      }
    } catch {
      // Stream torn down (cancel / completion) — expected.
    }
  })();

  ctx.reportStatus('running');
  let emitted = 0;
  try {
    for (let tick = 1; tick <= params.ticks && !ctx.cancelSignal.aborted; tick += 1) {
      // Park while paused. This is the real suspension the framework's
      // status-only PauseTask does not provide: no progress is emitted and
      // the tick counter does not advance until we resume (or cancel).
      while (state.paused && !ctx.cancelSignal.aborted) {
        await sleepMs(TICK_MS, ctx.cancelSignal);
      }
      if (ctx.cancelSignal.aborted) break;

      try {
        control.write({ tick, state: 'running' });
      } catch (err) {
        // The SDK force-ends the stream on fatal status / PAM revocation / a
        // teardown racing the cancel check; a write then throws. Break rather
        // than rethrow so cancel still converges on `canceled` (mirrors Python).
        if (isEndedStreamError(err)) break;
        throw err;
      }
      emitted += 1;
      await sleepMs(TICK_MS, ctx.cancelSignal);
    }
  } catch (err) {
    if (!isAbortError(err)) throw err;
  }

  await control.end().catch(() => undefined);
  await reader;

  const completionReason = ctx.isCancelled ? 'canceled' : 'completed';
  log(`Work loop ${completionReason} (${emitted} ticks, paused=${state.paused || state.resumed})`);
  ctx.reportStatus(completionReason);

  return {
    artifacts: [{
      data: JSON.stringify(
        { kind: 'pipe', ticks: emitted, paused: state.paused, resumed: state.resumed, completionReason },
        null,
        2,
      ),
      mimeType: 'application/json',
    }],
  };
}

const TICK_MS = 500;
const ABORT_MESSAGE = 'aborted';
const DEFAULT_TICKS = 20;
const MAX_TICKS = 200;

interface Params {
  ticks: number;
  failOnce: boolean;
}

function parseParams(parts: unknown[] | undefined): Params {
  let ticks = DEFAULT_TICKS;
  let failOnce = false;
  for (const part of parts ?? []) {
    const record = coerceRecord(part);
    if (!record) continue;
    if (typeof record.ticks === 'number' && Number.isFinite(record.ticks)) {
      ticks = Math.min(MAX_TICKS, Math.max(1, Math.trunc(record.ticks)));
    }
    if (typeof record.failOnce === 'boolean') {
      failOnce = record.failOnce;
    }
  }
  return { ticks, failOnce };
}

function coerceRecord(part: unknown): Record<string, unknown> | undefined {
  if (part === null || typeof part !== 'object' || Array.isArray(part)) {
    return undefined;
  }
  const record = part as Record<string, unknown>;
  if (typeof record.text === 'string') {
    try {
      const parsed = JSON.parse(record.text);
      if (parsed !== null && typeof parsed === 'object' && !Array.isArray(parsed)) {
        return parsed as Record<string, unknown>;
      }
    } catch {
      // text is not JSON — fall through to the raw record
    }
  }
  return record;
}

function sleepMs(ms: number, signal: AbortSignal): Promise<void> {
  return new Promise((resolve, reject) => {
    const timer = setTimeout(() => {
      signal.removeEventListener('abort', onAbort);
      resolve();
    }, ms);

    const onAbort = () => {
      clearTimeout(timer);
      signal.removeEventListener('abort', onAbort);
      reject(new Error(ABORT_MESSAGE));
    };

    signal.addEventListener('abort', onAbort, { once: true });
  });
}

function isAbortError(err: unknown): boolean {
  return err instanceof Error && err.message === ABORT_MESSAGE;
}

function isEndedStreamError(err: unknown): boolean {
  return err instanceof Error && err.message === 'Cannot write to an ended stream';
}

File: examples/python/lifecycle/handler.py
"""
lifecycle handler (Python) — mirrors the Node lifecycle example.

Teaches the boundary between task-lifecycle ops the framework backs
first-class and ops a provider composes from primitives:

  - cancel  (built-in)           cooperative: the SDK sets ctx.cancel_event
                                 and the loop stops -> terminal ``canceled``.
  - pause   (provider-composed)  NOT a handler hook — the framework's
  - resume  (provider-composed)  PauseTask/ResumeTask publish a status event
                                 only and never reach handler code. So we
                                 build real work-suspension: read
                                 { "ctrl": "pause" | "resume" } on a bidi
                                 control stream and park the work loop.
  - retry   (consumer-composed)  NOT done here — task state is in-memory only
                                 (SDK_CONTRACT §17), so retry is a consumer
                                 resubmit. This handler stays idempotent and
                                 honors a ``failOnce`` flag so the consumer can
                                 drive a failed->completed retry.

Task-kind split:
  - pipe:    long-running work loop with pause/resume/cancel (no auto-terminal).
  - request: single-shot; raises when ``failOnce`` is set (drives the retry demo).
"""

from __future__ import annotations

import json
import math
import threading
from typing import Any, Dict, List, Optional

from blocks_network.types import StartTaskMessage, TaskContext

TICK_SECONDS = 0.5
DEFAULT_TICKS = 20
MAX_TICKS = 200


def handler(task: StartTaskMessage, ctx: Optional[TaskContext] = None) -> Dict[str, Any]:
    params = _parse_params(task.request_parts)
    is_pipe = not task.task_kind or task.task_kind == "pipe"
    if not is_pipe:
        return _run_request(task, params)
    return _run_pipe(task, params, ctx)


# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------


def _log(msg: str) -> None:
    print(f"[lifecycle] {msg}")


def _artifact(payload: Dict[str, Any]) -> Dict[str, Any]:
    return {
        "artifacts": [{
            "data": json.dumps(payload, indent=2),
            "mimeType": "application/json",
        }],
    }


def _run_request(task: StartTaskMessage, params: Dict[str, Any]) -> Dict[str, Any]:
    """Request path: stateless fail-once retry demo.

    The consumer sends failOnce on attempt 1 (-> terminal ``failed``) and
    resubmits with a fresh idempotencyKey and no flag (-> ``completed``). No
    persisted "have I run before?" state is needed.
    """
    if params["fail_once"]:
        _log(f"Task {task.task_id}: failOnce set — raising to produce a failed terminal")
        raise RuntimeError("lifecycle: simulated first-attempt failure (failOnce)")
    _log(f"Task {task.task_id}: request completed")
    return _artifact({"kind": "request", "completionReason": "completed"})


def _run_pipe(
    task: StartTaskMessage, params: Dict[str, Any], ctx: Optional[TaskContext]
) -> Dict[str, Any]:
    """Pipe path: pause / resume / cancel work loop over a bidi control stream."""
    # Raise (not return) on a missing stream: a pipe handler's voluntary return
    # publishes NO terminal, so returning an error artifact would leave the task
    # lingering until duration expiry and then terminal as ``completed``. A raise
    # publishes an immediate, correct ``failed`` — matching the request path.
    if ctx is None or not ctx.has_stream:
        raise RuntimeError("lifecycle pipe task requires a negotiated stream")

    _log(f"Task {task.task_id}: starting work loop (up to {params['ticks']} ticks)")
    control = ctx.create_stream(declared_stream="control", direction="bidirectional", format="events")

    # `paused` is the single source of truth the work loop parks on; `resumed`
    # records that we came back from a pause.
    state = {"paused": False, "resumed": False}
    reader = threading.Thread(
        target=_read_control, args=(control, ctx, state), name="lifecycle-control", daemon=True
    )
    reader.start()

    emitted = _work_loop(control, ctx, params, state)

    try:
        control.end()
    except RuntimeError:
        pass  # already ended by SDK shutdown
    reader.join(timeout=2.0)

    completion_reason = "canceled" if ctx.is_cancelled else "completed"
    _log(f"Work loop {completion_reason} ({emitted} ticks, paused={state['paused'] or state['resumed']})")
    ctx.report_status(completion_reason)

    return _artifact(
        {
            "kind": "pipe",
            "ticks": emitted,
            "paused": state["paused"],
            "resumed": state["resumed"],
            "completionReason": completion_reason,
        }
    )


def _read_control(control: Any, ctx: TaskContext, state: Dict[str, bool]) -> None:
    """Read control messages on a background thread so the work loop never
    blocks waiting on input."""
    try:
        for event in control.events():
            ctrl = event.get("ctrl") if isinstance(event, dict) else None
            if ctrl == "pause" and not state["paused"]:
                state["paused"] = True
                _log("control: pause — parking work loop")
                ctx.report_status("paused")
            elif ctrl == "resume" and state["paused"]:
                state["paused"] = False
                state["resumed"] = True
                _log("control: resume — continuing work loop")
                ctx.report_status("running")
    except Exception as exc:  # noqa: BLE001 — stream torn down on cancel/completion is expected
        _log(f"control reader stopped: {exc}")


def _work_loop(control: Any, ctx: TaskContext, params: Dict[str, Any], state: Dict[str, bool]) -> int:
    ctx.report_status("running")
    emitted = 0
    tick = 0
    while tick < params["ticks"] and not ctx.is_cancelled:
        # Park while paused. This is the real suspension the framework's
        # status-only PauseTask does not provide: no progress is emitted and
        # the tick counter does not advance until we resume (or cancel).
        while state["paused"] and not ctx.is_cancelled:
            _sleep_or_cancel(TICK_SECONDS, ctx)
        if ctx.is_cancelled:
            break

        tick += 1
        try:
            control.write({"tick": tick, "state": "running"})
        except RuntimeError:
            break  # stream ended by SDK shutdown
        emitted += 1
        _sleep_or_cancel(TICK_SECONDS, ctx)
    return emitted


def _sleep_or_cancel(seconds: float, ctx: TaskContext) -> None:
    """Sleep up to ``seconds``, returning early if cancellation is requested."""
    # cancel_event.wait() returns True as soon as the event is set, so a
    # cancel interrupts the sleep immediately rather than after the full tick.
    ctx.cancel_event.wait(timeout=seconds)


def _parse_params(parts: Optional[List[Any]]) -> Dict[str, Any]:
    ticks = DEFAULT_TICKS
    fail_once = False
    for part in parts or []:
        content = _parse_part_content(part)
        value = content.get("ticks")
        # `json.loads` accepts NaN/Infinity literals; guard so a non-finite
        # `ticks` falls back to the default instead of crashing `int()` (Node
        # guards with `Number.isFinite`).
        if isinstance(value, (int, float)) and not isinstance(value, bool) and math.isfinite(value):
            ticks = min(MAX_TICKS, max(1, int(value)))
        flag = content.get("failOnce")
        if isinstance(flag, bool):
            fail_once = flag
    return {"ticks": ticks, "fail_once": fail_once}


def _parse_part_content(part: Any) -> Dict[str, Any]:
    text = getattr(part, "text", None) if not isinstance(part, dict) else part.get("text")
    if isinstance(text, str):
        try:
            parsed = json.loads(text)
            if isinstance(parsed, dict):
                return parsed
        except (json.JSONDecodeError, ValueError):
            pass
    if isinstance(part, dict):
        return part
    return {}
